File: src/main/java/com/chatchatabc/jpademojava/application/rest/UserController.java
package com.chatchatabc.jpademojava.application.rest;

import com.chatchatabc.jpademojava.application.dto.ErrorContent;
import com.chatchatabc.jpademojava.application.dto.user.UserPasswordUpdateRequest;
import com.chatchatabc.jpademojava.application.dto.user.UserProfileResponse;
import com.chatchatabc.jpademojava.application.dto.user.UserProfileUpdateRequest;
import com.chatchatabc.jpademojava.domain.model.User;
import com.chatchatabc.jpademojava.domain.repository.UserRepository;
import com.chatchatabc.jpademojava.domain.service.UserService;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.Optional;

@RestController
@RequestMapping("/api/user")
public class UserController {
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private UserService userService;

    private final ModelMapper modelMapper = new ModelMapper();

    /**
     * Get user profile
     *
     * @return
     */
    @GetMapping("/profile")
    public ResponseEntity<UserProfileResponse> profile() {
        try {
            // Get id from security context
            User principal = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
            Optional<User> user = userRepository.findById(principal.getId());
            if (user.isEmpty()) {
                throw new Exception("User not found");
            }
            return ResponseEntity.ok(new UserProfileResponse(user.get(), null));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(new UserProfileResponse(null, new ErrorContent("User Profile Error", e.getMessage())));
        }
    }

    /**
     * View profile of other user
     *
     * @param username
     * @return
     */
    @GetMapping("/profile/view/{username}")
    public ResponseEntity<UserProfileResponse> viewProfile(
            @PathVariable String username
    ) {
        try {
            Optional<User> user = userRepository.findByUsername(username);
            if (user.isEmpty()) {
                throw new Exception("User not found");
            }
            return ResponseEntity.ok(new UserProfileResponse(user.get(), null));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(new UserProfileResponse(null, new ErrorContent("User Profile Error", e.getMessage())));
        }
    }

    /**
     * Update user profile
     *
     * @param request
     * @return
     */
    @PutMapping("/profile/update")
    public ResponseEntity<UserProfileResponse> updateProfile(
            @RequestBody UserProfileUpdateRequest request
    ) {
        try {
            // Get id from security context
            User principal = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
            User newUserInfo = modelMapper.map(request, User.class);
            User user = userService.update(principal.getId(), newUserInfo);
            return ResponseEntity.ok(new UserProfileResponse(user, null));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(new UserProfileResponse(null, new ErrorContent("User Profile Error", e.getMessage())));
        }
    }

    /**
     * Update user password
     *
     * @param request
     * @return
     */
    @PutMapping("/profile/change-password")
    public ResponseEntity<UserProfileResponse> updatePassword(
            @RequestBody UserPasswordUpdateRequest request
    ) {
        try {
            // Get id from security context
            User principal = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
            User user = userService.updatePassword(principal.getId(), request.getOldPassword(), request.getNewPassword());
            return ResponseEntity.ok(new UserProfileResponse(user, null));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(new UserProfileResponse(null, new ErrorContent("User Profile Error", e.getMessage())));
        }
    }
}

File: src/main/kotlin/com/chatchatabc/jpademo/application/rest/UserController.kt
package com.chatchatabc.jpademo.application.rest

import com.chatchatabc.jpademo.application.dto.ErrorContent
import com.chatchatabc.jpademo.application.dto.user.UserPasswordUpdateRequest
import com.chatchatabc.jpademo.application.dto.user.UserProfileResponse
import com.chatchatabc.jpademo.application.dto.user.UserProfileUpdateRequest
import com.chatchatabc.jpademo.domain.model.User
import com.chatchatabc.jpademo.domain.repository.UserRepository
import com.chatchatabc.jpademo.domain.service.UserService
import org.springframework.http.ResponseEntity
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/user")
class UserController(
    private val userRepository: UserRepository,
    private val userService: UserService
) {
    /**
     * User Profile
     */
    @GetMapping("/profile")
    fun profile(): ResponseEntity<UserProfileResponse> {
        return try {
            // Get id from security context
            val principal = SecurityContextHolder.getContext().authentication.principal as User
            val user = userRepository.findById(principal.id)
            if (user.isEmpty) {
                throw Exception("User not found")
            }
            ResponseEntity.ok(UserProfileResponse(user.get(), null))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(UserProfileResponse(null, ErrorContent("User Profile Error", e.message)))
        }
    }

    /**
     * View profile of other user
     */
    @GetMapping("/profile/view/{username}")
    fun viewProfile(
        @PathVariable username: String
    ): ResponseEntity<UserProfileResponse> {
        return try {
            val user = userRepository.findByUsername(username)
            if (user.isEmpty) {
                throw Exception("User not found")
            }
            ResponseEntity.ok(UserProfileResponse(user.get(), null))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(UserProfileResponse(null, ErrorContent("User Profile Error", e.message)))
        }
    }

    /**
     * Update User Profile
     */
    @PutMapping("/profile/update")
    fun updateProfile(
        @RequestBody request: UserProfileUpdateRequest
    ): ResponseEntity<UserProfileResponse> {
        return try {
            // Get id from security context
            val principal = SecurityContextHolder.getContext().authentication.principal as User
            val user = userService.update(principal.id, request)
            ResponseEntity.ok(UserProfileResponse(user, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(UserProfileResponse(null, ErrorContent("User Profile Error", e.message)))
        }
    }

    /**
     * Update User Password
     */
    @PutMapping("/profile/change-password")
    fun updatePassword(
        @RequestBody request: UserPasswordUpdateRequest
    ): ResponseEntity<UserProfileResponse> {
        return try {
            // Get id from security context
            val principal = SecurityContextHolder.getContext().authentication.principal as User
            val user = userService.updatePassword(principal.id, request.oldPassword, request.newPassword)
            ResponseEntity.ok(UserProfileResponse(user, null))
        } catch (e: Exception) {
            ResponseEntity.badRequest()
                .body(UserProfileResponse(null, ErrorContent("User Change Password Error", e.message)))
        }
    }
}